binding.cc

```diff
@@ -952,28 +995,46 @@ namespace zmq {
     if(zmq_socket_monitor(socket->socket_, addr, ZMQ_EVENT_ALL) != -1) {
       socket->monitor_socket_ = zmq_socket (context->context_, ZMQ_PAIR);
       zmq_connect (socket->monitor_socket_, addr);
-
+      socket->timer_interval_ = timer_interval;
+      socket->num_of_events_ = num_of_events;
       socket->monitor_handle_ = new uv_timer_t;
-
       socket->monitor_handle_->data = socket;
 
       uv_timer_init(uv_default_loop(), socket->monitor_handle_);
-      uv_timer_start(socket->monitor_handle_, reinterpret_cast<uv_timer_cb>(Socket::UV_MonitorCallback), timer_interval, timer_interval);
+      uv_timer_start(socket->monitor_handle_, reinterpret_cast<uv_timer_cb>(Socket::UV_MonitorCallback), timer_interval, 0);
     }
 
     NanReturnUndefined();
   }
 
+  void
+  Socket::Unmonitor() {
+    // Make sure we are monitoring
+    if (this->monitor_socket_ == NULL) {
+      return;
+    }
+
+    // Passing NULL as addr will tell zmq to stop monitor
+    zmq_socket_monitor(this->socket_, NULL, ZMQ_EVENT_ALL);
+
+    // Close the monitor socket and stop timer
+    if (zmq_close(this->monitor_socket_) < 0)
+      throw std::runtime_error(ErrorMessage());
```

Comment:
Emit an error event instead.

On June 3, 2015 7:02:27 PM EEST, Sean Mathews <notifications@github.com> wrote:
>>      }
>>  
>>      NanReturnUndefined();
>>    }
>>  
>> +  void
>> +  Socket::Unmonitor() {
>> +    // Make sure we are monitoring
>> +    if (this->monitor_socket_ == NULL) {
>> +      return;
>> +    }
>> +
>> +    // Passing NULL as addr will tell zmq to stop monitor
>> +    zmq_socket_monitor(this->socket_, NULL, ZMQ_EVENT_ALL);
>> +
>> +    // Close the monitor socket and stop timer
>> +    if (zmq_close(this->monitor_socket_) < 0)
>> +      throw std::runtime_error(ErrorMessage());
>
>I have been digging and digging and found no good ideas on handling an
>exception inside of uv_timer effectively. A little frustrating. I have
>a few ideas I want to throw out but again this is a desperate approach
>and may be non standard. What if we add an optional argument to
>monitor() as a function to call upon exception?
>
>req.monitor(10, 0, function(e) {
>  /* handle error or events from uv_timer with args in e */  
>});
>
>this would require us to save the context and the function pointer and
>I am sure close attention needs to be paid to not create any leaks or
>gc problems.
>
>Thoughts?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/JustinTulloss/zeromq.node/pull/424/files#r31639144